engine/grid.py

```diff
@@ -300,73 +307,164 @@ def find_border_nodes(self):
             TERMINATE = 2
-        rows = self.nodes.shape[0]
+        class Direction(Enum):
+            RIGHT = 1
+            LEFT = 2
+
+        class Orientation(Enum):
+            CW = 1
+            CCW = 2
+
+        def plot_circle(start_pos, end_pos, center, orientation):
+            """
+            Returns a circle of nodes starting from the start_pos, going at orientation orientation, and ending at the
+            end_pos with center center.
+
+            Arguments:
+                start_pos: float tuple representing the starting point of the circle being plotted
+                end_pos: float tuple representing the ending point of the circle being plotted
+                center: float tuple representing the center point of the circle being plotted
+                orientation: orientation that the nodes are being plotted from, starting with the start_pos and ending
+                    at the end_pos
+            """
+            r = math.hypot(float(start_pos[0]) - center[0], float(start_pos[1]) - center[1])
+            theta_init = math.atan2(start_pos[1]-center[1], start_pos[0]-center[0])
+            theta_end = math.atan2(end_pos[1]-center[1], end_pos[0]-center[0])
+            circle_plt = []
+            if theta_end < theta_init:
+                theta_end = theta_end + 2*math.pi
+            if orientation == Orientation.CCW:
+                theta = theta_init
+                while theta < theta_end:
+                    circle_plt.append((r*math.cos(theta)+center[0], r*math.sin(theta)+center[1]))
+                    theta = theta + math.pi/12
+            elif orientation == Orientation.CW:
+                theta = theta_end
+                while theta > theta_init:
+                    circle_plt.append((r*math.cos(theta)+center[0], r*math.sin(theta)+center[1]))
+                    theta = theta - math.pi/12
+            else:
+                raise Exception("invalid orientation")
+            return circle_plt
+
         phase = WaypointPhase.DOWN
         curr_pos = self.leftmost_node_pos
+        direction = Direction.RIGHT
         waypoints = []
         waypoints.append(curr_pos)
         while (phase != WaypointPhase.TERMINATE):
-            if (phase == WaypointPhase.DOWN):
-                new_pos = (curr_pos[0]+1, curr_pos[1])
-                if curr_pos[0]+1 == rows and self.nodes[new_pos].is_active:
+            if direction == Direction.LEFT:
+                new_pos = (curr_pos[0]-1,curr_pos[1])
+                if self.nodes[new_pos].is_active:
                     waypoints.append(new_pos)
-                    right_pos = self.bottom_rightmost_node(new_pos)
-                    if right_pos is not None:
-                        waypoints.append(right_pos)
-                        curr_pos = right_pos
-                        phase = WaypointPhase.DOWN
+                    curr_pos = new_pos
+                else:
+                    left_pos = self.bottom_rightmost_node(new_pos)
```

Comment:
does bottom_rightmost_node of the new_position return the left_pos? I'm confused by the naming and what bottom_rightmost_node does

Reply:
also, it looks like bottom_rightmost_node return the right-most activated node of the row directly above the current row, is that right?

Reply:
Yes, that is right. I have combined bottom_rightmost_node and bottom_rightmost_node and renamed the new function nextrow_sidemost_node